### PR TITLE
Added Vulture dead code checker

### LINKED ISSUE #5649: add check for dead code
## 🚀 Feature

check if PR contains unused code or if the particular change dead code in not introduced somewhere else in the codebase

### Motivation

remove all unused code, keep the codebase healthy

### Alternatives

one toll I found - https://github.com/jendrikseipp/vulture
but let's check some alternatives

### PR DESCRIPTION
**Pull request by aniketmaurya**
## What does this PR do?
Resolves #5649 
I have integrated [Vulture](https://github.com/jendrikseipp/vulture) to CI for dead code checking.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: .github/workflows/code-formatting.yml [hunk 1]
```diff
@@ -74,3 +74,24 @@ jobs:
       - name: mypy check
         run: |
           mypy
+
+  python-vulture:
```
Comment (anshajk) [inline .github/workflows/code-formatting.yml:78]: I've been following the issue for which this PR has been raised and I wonder if it makes sense to have this as a separate workflow instead of adding it in the code formatting workflow.

Comment (Borda) [inline .github/workflows/code-formatting.yml:78]:
```suggestion
  dead-code-vulture:
```

File: .github/workflows/code-formatting.yml [hunk 2]
```diff
@@ -74,3 +74,24 @@ jobs:
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.7
+
+      - name: Install dependencies
+        run: |
+          pip install vulture
```
Comment (Borda) [inline .github/workflows/code-formatting.yml:93]:
```suggestion
          pip install vulture
          pip list
```

File: pyproject.toml [hunk 3]
```diff
@@ -53,3 +53,10 @@ profile = "black"
 line_length = 120
 force_sort_within_sections = "False"
 order_by_type = "False"
+
+[tool.vulture]
+make_whitelist = true
+min_confidence = 100
+paths = ["pytorch_lightning"]
+sort_by_size = true
+verbose = false
```
Comment (Borda) [inline pyproject.toml:62]:
```suggestion
make_whitelist = true
min_confidence = 95
paths = ["pytorch_lightning"]
sort_by_size = true
```

**Review by Borda — COMMENTED** [review]
does it report any issue now?

**Comment by Borda** [comment]
@aniketmaurya thx for the quick PR, mind rebase it on feat 1.2

**Comment by aniketmaurya** [comment]
> does it report any issue now?

Hi @Borda, You mean in the GitHub actions? Yes, there are few dead code blocks detected by it.

**Comment by Borda** [comment]
> > does it report any issue now?
> 
> Hi @Borda, You mean in the GitHub actions? Yes, there are few dead code blocks detected by it.

ok, soo it shall be fixed within the PR :]
but fist can you rebase it in `release/1.2-dev`?

**Comment by aniketmaurya** [comment]
> > > does it report any issue now?
> > 
> > 
> > Hi @Borda, You mean in the GitHub actions? Yes, there are few dead code blocks detected by it.
> 
> ok, soo it shall be fixed within the PR :]
> but fist can you rebase it in `release/1.2-dev`?

@Borda  I think I messed up something during hard reset to 1.2 :(

**Comment by Borda** [comment]
@aniketmaurya fixed :]

**Comment by aniketmaurya** [comment]
> @aniketmaurya fixed :] can you fix the code so the check pass :]



> @aniketmaurya fixed :] can you fix the code so the check pass :]

There are a total of 37 unused variables detected in different modules. Should I check and remove them or add them to whitelist? I am not familiar with all of the modules.

**Comment by rohitgr7** [comment]
Hi @RJ722, mind review the vulture workflow CI here?

**Comment by awaelchli** [comment]
two dumb question because I do not know these tools very well:

how will this affect our workflow?
and will there be ugly `# vulture: ignore` in the code base as we are forced to do sometimes like` #type ignore` for mypy?

**Comment by Borda** [comment]
> and will there be ugly `# vulture: ignore` in the code base as we are forced to do sometimes like` #type ignore` for mypy?

I think that all these ignores are just temporary hacks and all shall be resolved properly

File: .github/workflows/code-formatting.yml [hunk 4]
```diff
@@ -74,3 +74,25 @@ jobs:
...
+          pip list
+
+      - name: Run checking
```
Comment (RJ722) [inline .github/workflows/code-formatting.yml:96]: Nitpick:

```suggestion
      - name: Check for dead code with Vulture
```

**Review by RJ722 — COMMENTED** [review]
I see that Vulture is failing right now. And people are not willing to annotate false positives with `#noqa` comments. The other way round this issue is to use a whitelist, which is a Python file enlisting the names of all the unused objects reported. More info [here](https://github.com/jendrikseipp/vulture#handling-false-positives). To make a whitelist right now, you can run `vulture pytorch_lightning --make-whitelist`.

Otherwise, the changes look good! I've suggested a minor nitpick, but it's entirely upto you.

**Comment by Borda** [comment]
@aniketmaurya how is it going here? :]

**Comment by aniketmaurya** [comment]
> @aniketmaurya how is it going here? :]

I'll push the changes by either by today evening or tomorrow morning.

**Comment by aniketmaurya** [comment]
Hi @Borda, I have renamed the unused variables to **_** and in case of multiple unused args in the same function replaced with **__**

File: tests/models/test_hparams.py [hunk 5]
```diff
@@ -60,7 +60,7 @@ class SaveHparamsDecoratedModel(BoringModel):
     """ Tests that a model can take an object """
     @decorate
     @decorate
-    def __init__(self, hparams, *my_args, **my_kwargs):
+    def __init__(self, hparams, *_, **__):
```
Comment (awaelchli) [inline tests/models/test_hparams.py:63]: The argument names for the hyperparameter tests here in this file needs to stay like it is, please don't change it.

File: pl_examples/domain_templates/reinforce_learn_Qnet.py [hunk 6]
```diff
@@ -323,7 +323,7 @@ def dqn_mse_loss(self, batch: Tuple[torch.Tensor, torch.Tensor]) -> torch.Tensor
 
         return nn.MSELoss()(state_action_values, expected_state_action_values)
 
-    def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor], nb_batch) -> OrderedDict:
+    def training_step(self, batch: Tuple[torch.Tensor, torch.Tensor], _) -> OrderedDict:
```
Comment (awaelchli) [inline pl_examples/domain_templates/reinforce_learn_Qnet.py:326]: it should simply be called batch_idx

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1133,7 +1133,7 @@ def training_step(...):
         self.trainer.train_loop.backward(loss, optimizer, -1, *args, **kwargs)
         self._running_manual_backward = False
 
-    def backward(self, loss: Tensor, optimizer: Optimizer, optimizer_idx: int, *args, **kwargs) -> None:
+    def backward(self, loss: Tensor, optimizer: Optimizer, _: int, *args, **kwargs) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1136]: don't rename this. this hook will be overridden by user

**Review by awaelchli — CHANGES_REQUESTED** [review]
I do not understand most changes in this PR. By replacing function arguments with underscore the code becomes less readable. Also the changes in the hparams test are wrong, and the tests will fail.

File: tests/models/test_hparams.py (hunk 5 above)
Comment (Borda) [reply tests/models/test_hparams.py:63]: do you have any idea how to deal with inherited API which is not used in this class?
if we would need to skip/ignore all then this Volture does not have much sense...
@aniketmaurya mind check if there is an option to deal in inherited API?

Comment (aniketmaurya) [reply tests/models/test_hparams.py:63]: @borda Another way is that we can either rename the variable to start with underscore (**_kwargs), Vulture will ignore these variables. Or we can add these to whitelist.

Comment (aniketmaurya) [reply tests/models/test_hparams.py:63]: I think training step variables should not be renamed. Then maybe we can ignore these names by maintaining [`ignore_names`](https://github.com/jendrikseipp/vulture#configuration) list.
We can keep something like this in `pyproject.toml`:
`ignore_names = ["batch_idx", "optimizer_idx", "batch"]`

And for variables like `args` and `kwargs` we can use underscore notation `*_args` and `**_kwargs`.

Comment (Borda) [reply tests/models/test_hparams.py:63]: @awaelchli how do you feel about `*_args` and `**_kwargs`?

Comment (awaelchli) [reply tests/models/test_hparams.py:63]: no matter what the renaming is, the tests must pass and they currently don't, because in the tests it matters what the var names are even if the tool says it is unused, it is actually used!
the dead code checker probably only looks at the static source files, but it can not determine such logic here.

**Comment by aniketmaurya** [comment]
Hi @Borda, I have added `ignore_names` to `pyproject.toml` and the vulture check has passed. But I don't understand why import formatting test is failing. On my forked repo is has passed https://github.com/aniketmaurya/pytorch-lightning/runs/1823785683

**Review by tchaton — APPROVED** [review]

**Comment by edenlightning** [comment]
@aniketmaurya mind resolving conflicts?

**Comment by aniketmaurya** [comment]
Sure will do @edenlightning

**Comment by aniketmaurya** [comment]
lots of dead code is being detected in `pytorch_lightning/metrics` module. Is it because we are moving to torchmetrics?
cc: @Borda

**Comment by Borda** [comment]
> lots of dead code is being detected in `pytorch_lightning/metrics` module. Is it because we are moving to torchmetrics?
> cc: @Borda

yes, pls ignore all inside metrics :]

**Comment by aniketmaurya** [comment]
> @aniketmaurya mind resolving conflicts?

Done @edenlightning !

**Comment by carmocca** [comment]
I'm not sure we should integrate this checker into Lightning, considering we have a lot of dynamism in the way we call hooks so this checker will often produce false positives.

The number of items you already had to whitelist makes me think this is not that useful for this project.

File: pyproject.toml [hunk 8]
```diff
@@ -16,3 +16,26 @@ profile = "black"
 line_length = 120
 force_sort_within_sections = "False"
 order_by_type = "False"
+
+[tool.vulture]
+exclude = ['pytorch_lightning/metrics']
+make_whitelist = true
+min_confidence = 95
+paths = ["pytorch_lightning"]
+ignore_names = [
```
Comment (Borda) [inline pyproject.toml:25]: why are these ignored, it is a kind of todo to be fixed and removed from the list?

Comment (aniketmaurya) [reply pyproject.toml:25]: yes, a lot of these variables were giving false positive.

**Comment by ethanwharris** [comment]
I agree with @carmocca, seems there are a lot of false positives (and a long whitelist) so I'm not sure there is much benefit to adding this.

**Comment by Borda** [comment]
I do not have any preference on used tool, but still thinking that we shall address dead code in our codebase...

File: .github/workflows/code-formatting.yml [hunk 9]
```diff
@@ -74,3 +74,25 @@ jobs:
...
+      - name: Checkout
+        uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: 3.7
+
+      - name: Install dependencies
+        run: |
+          pip install vulture
+          pip list
```
Comment (Borda) [inline .github/workflows/code-formatting.yml:94]:
```suggestion
        run: pip install vulture && pip list
```

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py [hunk 10]
```diff
@@ -90,7 +90,7 @@ def __enter__(self) -> None:
 
                 self._handles[module_name] = [pre_forward_handle, post_forward_handle]
 
-    def __exit__(self, exc_type: Any, exc_value: Any, traceback: Any) -> None:
+    def __exit__(self, type: Any, value: Any, traceback: Any) -> None:
```
Comment (awaelchli) [inline pytorch_lightning/profiler/pytorch.py:93]: type is a built-in keyword. 
better to avoid name clash here

File: .github/workflows/code-formatting.yml [hunk 11]
```diff
@@ -74,3 +74,23 @@ jobs:
       - name: mypy check
         run: |
           mypy
+
+  dead-code-vulture:
```
Comment (awaelchli) [inline .github/workflows/code-formatting.yml:78]: if it fails, will it block merging a PR?
@Borda

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/profiler/pytorch.py (hunk 10 above)
Comment (ethanwharris) [reply pytorch_lightning/profiler/pytorch.py:93]: Vulture picks up `exc_type` because it doesn't match the parent signature. We have a few options:
- leave like this, matches the parent
- convert to `_`, `__`, `___` - standard for unused variables
- add to whitelist (already a bit big)

File: .github/workflows/code-formatting.yml (hunk 11 above)
Comment (ethanwharris) [reply .github/workflows/code-formatting.yml:78]: Shouldn't do. Comes up as a serperate check that isn't required for merge.

Comment (Borda) [reply .github/workflows/code-formatting.yml:78]: No, it is not required check
